src/cunumeric/scan/scan_local_util.h

```diff
@@ -0,0 +1,72 @@
+/* Copyright 2021-2022 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#pragma once
+
+#include "cunumeric/cunumeric.h"
+
+#include <thrust/functional.h>
+
+namespace cunumeric {
+
+enum class ScanCode : int {
+  PROD = CUNUMERIC_SCAN_PROD,
+  SUM  = CUNUMERIC_SCAN_SUM,
+};
+
+template <typename Functor, typename... Fnargs>
+constexpr decltype(auto) op_dispatch(ScanCode op_code, bool nan0, Functor f, Fnargs&&... args)
+{
+  switch (op_code) {
+    case ScanCode::PROD:
+      switch (nan0) {
+        case true:
+          return f.template operator()<ScanCode::PROD, true>(std::forward<Fnargs>(args)...);
+        case false:
+          return f.template operator()<ScanCode::PROD, false>(std::forward<Fnargs>(args)...);
+      }
+    case ScanCode::SUM:
+      switch (nan0) {
+        case true: return f.template operator()<ScanCode::SUM, true>(std::forward<Fnargs>(args)...);
+        case false:
+          return f.template operator()<ScanCode::SUM, false>(std::forward<Fnargs>(args)...);
+      }
+    default: break;
+  }
+  assert(false);
+  return f.template operator()<ScanCode::SUM, false>(std::forward<Fnargs>(args)...);
+}
+
+// RRRR not sure I fully understand these?
+
+template <ScanCode OP_CODE, legate::LegateTypeCode CODE>
+struct ScanOp {
```

Comment:
I added `nan_null` to the base struct since similar to valid in the other OP templatized routines, it exists in all instances. Otherwise it should function correctly if it's only included in the instances.
Do you want me to remove it?

Reply:
yes, the default value for `valid` controls template instantiations, but `nan_null` is only used with individual scan operations, so doesn't need the default value to be defined.

Reply:
1db1596bbe66e90f6ad73d010418dd8d7005bbad